include/revng-c/mlir/Dialect/Clift/IR/CliftStorage.h

```diff
@@ -12,166 +14,109 @@
 
 namespace mlir::clift {
 
-struct StructTypeStorage : public mlir::AttributeStorage {
-public:
-  // Some internal mechanism of mlir have access to the key of the storage
-  // instead of the entire storage or the key using that storage. For that
-  // reason it is best to throw everything related to the type inside the key,
-  // and then override operators to pretend they don't know about the non key
-  // fields. That way when it is needed one can access everything.
-  struct Key {
+template<typename StorageT, typename ValueT>
+class ClassTypeStorage : public mlir::AttributeStorage {
+  struct KeyBase {
     uint64_t ID;
-    llvm::StringRef name;
-    uint64_t Size;
```

Comment:
Ok so the `Size` cannot be part of the `StructTypeStorage` now? How so? It must have to do with the fact that you allow it to be added afterwards when you complete the type. Is that the case?

Reply:
Or is it just to allow to CRTP all the `*TypeStorage` stuff?

Reply:
`StructTypeStorage` and `UnionTypeStorage` are almost the same, the only difference being the additional `uint64_t Size` member in the `StructTypeStorage` key type. That's why I factored out the common parts into `ClassTypeStorage`.
`uint64_t Size` is still there: https://github.com/revng/revng-c/pull/1/files#diff-fa79273e9b99570adf298a58172247aa90dfa23f27e2c87a9b94c7fa0ff8c9d4R107

Reply:
:+1: